src/main/java/net/neoforged/neoforge/attachment/AttachmentSync.java

```diff
@@ -0,0 +1,251 @@
+                player.connection.send(packet);
+            }
+        }
+    }
+
+    public static void syncBlockEntityUpdate(BlockEntity blockEntity, AttachmentType<?> type) {
+        if (type.syncHandler == null || !(blockEntity.getLevel() instanceof ServerLevel serverLevel)) {
+            return;
+        }
+        syncUpdate(blockEntity, type, serverLevel.getChunkSource().chunkMap.getPlayers(new ChunkPos(blockEntity.getBlockPos()), false));
+    }
+
+    public static void syncChunkUpdate(LevelChunk chunk, AttachmentHolder.AsField holder, AttachmentType<?> type) {
+        if (type.syncHandler == null || !(chunk.getLevel() instanceof ServerLevel serverLevel)) {
+            return;
+        }
+        syncUpdate(holder, type, serverLevel.getChunkSource().chunkMap.getPlayers(chunk.getPos(), false));
+    }
+
+    public static void syncEntityUpdate(Entity entity, AttachmentType<?> type) {
+        if (type.syncHandler == null || !(entity.level() instanceof ServerLevel serverLevel)) {
+            return;
+        }
+        var players = serverLevel.getChunkSource().chunkMap.getPlayersWatching(entity);
+        if (entity instanceof ServerPlayer serverPlayer) {
+            // Players do not track themselves
+            var newPlayers = new ArrayList<ServerPlayer>(players.size() + 1);
+            newPlayers.addAll(players);
+            newPlayers.add(serverPlayer);
+            players = newPlayers;
+        }
+        syncUpdate(entity, type, players);
+    }
+
+    public static void syncLevelUpdate(ServerLevel level, AttachmentType<?> type) {
+        if (type.syncHandler == null) {
+            return;
+        }
+        syncUpdate(level, type, level.players());
+    }
+
+    /**
+     * Constructs a payload to sync all syncable attachments to a player, if any.
+     */
+    @Nullable
+    private static SyncAttachmentsPayload syncInitialAttachments(AttachmentHolder holder, ServerPlayer to) {
+        if (holder.attachments == null) {
+            return null;
+        }
+        boolean anySyncableAttachment = false;
+        for (var attachment : holder.attachments.keySet()) {
+            anySyncableAttachment = anySyncableAttachment | attachment.syncHandler != null;
+        }
+        if (!anySyncableAttachment) {
+            return null;
+        }
+        List<AttachmentType<?>> syncedTypes = new ArrayList<>();
+        var data = FriendlyByteBufUtil.writeCustomData(buf -> {
+            for (var entry : holder.attachments.entrySet()) {
+                AttachmentType<?> type = entry.getKey();
+                @SuppressWarnings("unchecked")
+                var syncHandler = (AttachmentSyncHandler<Object>) type.syncHandler;
+                if (syncHandler != null) {
+                    int indexBefore = buf.writerIndex();
+                    buf.writeBoolean(true);
+                    syncHandler.write(buf, entry.getValue(), true);
+                    if (indexBefore < buf.writerIndex()) {
```

Comment:
This check will always be true because `indexBefore` is captured before the "add" boolean is written to the buffer. You'd have to do something like this for it to actually do what you intend:
```java
if (syncHandler != null) {
    int indexBefore = buf.writerIndex();
    buf.writeBoolean(true);
    int indexBetween = buf.writerIndex();
    syncHandler.write(buf, entry.getValue(), true);
    if (indexBetween < buf.writerIndex()) {
        // Actually wrote something
        syncedTypes.add(type);
    } else {
        buf.writerIndex(indexBefore);
    }
}
```
The conversion to a byte array in `FriendlyByteBufUtil.writeCustomData()` will correctly drop the boolean if the last encountered attachment type didn't write anything and the writer index is therefore moved back.